### PR TITLE
Validations

### PR DESCRIPTION
**Pull request by rachitchaurasia99**
```
Validations

In product model add validations for following
  Permalink
   -should be unique
   -no special character and no space allowed
   -minimum 3 words, separated by hyphen 
  Price numericality validation should only run if price is present.
  Description - should be between 5 and 10 words
  Validate image_url format by using url:true(ActiveModel::EachValidator)
   In product model, price should be greater than discount price
   -using custom validator
   -without custom validator

LineItems model
Combination of product_id and cart_id should be unique. Means one product can be added only once in the cart. If cart with cart_id exists.

Add following validations in user model
Uniqueness of email
Email format
```

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,5 +1,6 @@
 class User < ApplicationRecord
   validates :name, presence: true, uniqueness: true
+  validates :email, uniqueness: true, format: {with: /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i}
```
Comment (apekshit-tewathia) [inline app/models/user.rb]: extract regex in a constant everywhere

Comment (apekshit-tewathia) [inline app/models/user.rb]: what would happen if i pass in an empty email. will this validation fail? should this validation fail?

File: app/models/line_item.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@ class LineItem < ApplicationRecord
   belongs_to :order, optional: true
   belongs_to :product
   belongs_to :cart, optional: true
+  validates :product, uniqueness: {scope: :cart}
```
Comment (apekshit-tewathia) [inline app/models/line_item.rb]: is there any difference if we used product_id and cart_id instead of product and cart here
Please read about allow_blank and allow_nil and tell me the difference and see if it can/should be used here

File: README.md [hunk 3]
```diff
@@ -1,24 +1,3 @@
-# README
+# readme
```
Comment (apekshit-tewathia) [inline README.md]: there should be no extra changes in the PR than the ones requested

File: app/models/product.rb [hunk 4]
```diff
@@ -1,8 +1,30 @@
 class Product < ApplicationRecord
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+  class PriceValidator < ActiveModel::Validator
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: let's move these validators to new files called price_validator and url_validator etc.

File: app/models/product.rb [hunk 5]
```diff
@@ -14,4 +36,13 @@ def ensure_not_referenced_by_any_line_item
       throw :abort
     end
    end
+
+   def hyphenated
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: these methods should not be needed. Let's do both of these using regex

File: app/models/product.rb [hunk 6]
```diff
@@ -1,8 +1,30 @@
...
+  end
+  # validates_with PriceValidator
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: { message: "can't be blank"}
   validates :title, uniqueness: true
   validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :permalink, uniqueness: true, format: {with: /[^[:alnum:]|-]/i}
+  validates :hyphenated, length: {minimum: 3, message: "Permalink must be of minimum 3 characters"}, allow_blank: true
+  validates :price, numericality: { greater_than_or_equal_to: 0.01 },if: :price
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: extract all values in constants

File: app/models/product.rb [hunk 7]
```diff
@@ -1,8 +1,30 @@
...
+      unless value =~ %r{\.(gif|jpg|png)\z}i
+        record.errors.add(attribute,
+          (options[:message] || "is not an approriate Image URL format")
+        )
+      end
+    end
+  end
+  # validates_with PriceValidator
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: { message: "can't be blank"}
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: what's the default message?

**Review by apekshit-tewathia — CHANGES_REQUESTED** [review]

File: app/models/product.rb (hunk 7 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: The default message is itself  "can't be blank" , so we can remove the message here

File: app/models/user.rb (hunk 1 above)
Comment (rachitchaurasia99) [reply app/models/user.rb]: done sir

File: app/models/product.rb (hunk 4 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done sir

File: app/models/product.rb (hunk 6 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done sir.

File: app/models/user.rb (hunk 1 above)
Comment (rachitchaurasia99) [reply app/models/user.rb]: Yes it will fail with an error message "Email is invalid" , but if we want to give an error for nil email then we can use `presence: true` before format

File: app/models/line_item.rb (hunk 2 above)
Comment (rachitchaurasia99) [reply app/models/line_item.rb]: `validates :product, uniqueness: {scope: :cart}` will also look for product_id and cart_id but its better to pass the attribute rather than whole object so we should use product_id and cart_id here.

`allow_nil` skips validation for nil values
`allow_blank` skips validation for nil and empty string values 
default value for both is 'false'

Here we don't need it because when we insert a nil or blank value then `belongs_to :product` will raise an error.

File: README.md (hunk 3 above)
Comment (apekshit-tewathia) [reply README.md]: Please remove this

File: app/controllers/concerns/url_validator.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class UrlValidator < ActiveModel::EachValidator
+    def validate_each(record, attribute, value)
+      unless value =~ IMAGE_URL_REGEX
+        record.errors.add(attribute,
+          (options[:message] || "is not an approriate Image URL format")
+        )
+      end
+    end
+  end
```
Comment (apekshit-tewathia) [inline app/controllers/concerns/url_validator.rb]: this newline symbol should never come in a PR. There is some setting in sublime/vscode that adds a new line at the end of file every time you save it. Please use that

File: app/models/product.rb (hunk 5 above)
Comment (apekshit-tewathia) [reply app/models/product.rb]: still not removed

File: README.md
Comment (apekshit-tewathia) [inline README.md:1]: why is this being deleted

File: app/models/product.rb [hunk 9]
```diff
@@ -1,17 +1,32 @@
...
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+
+  validates_with PriceValidator
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: true
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :image_url, allow_blank: true, format: { with: IMAGE_URL_REGEX, message: 'must be a URL for GIF, JPG or PNG image.'}
+  validates :permalink, uniqueness: true, format: { with: PERMALINK_REGEX, message: 'is not in valid format'}
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: error message not needed

File: app/models/product.rb [hunk 10]
```diff
@@ -1,17 +1,32 @@
...
+  validates :price, numericality: { greater_than_or_equal_to: 0.01 },if: :price
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: add space after comma.
find another way to do so

File: app/models/product.rb [hunk 11]
```diff
@@ -1,17 +1,32 @@
...
+  validates :price, numericality: {greater_than: :discount_price, message: "Discount Price can't be greater than Original Price"}
+  validates :words_in_description, length: { in: 5..10 }
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: do it using regex also

File: app/models/product.rb [hunk 12]
```diff
@@ -1,17 +1,32 @@
...
+
   # ensure that there are no line items referencing this product
    def ensure_not_referenced_by_any_line_item
     unless line_items.empty?
       errors.add(:base, 'Line Items present')
       throw :abort
     end
    end
+
+   def words_in_description
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: avoid having methods like thes

File: app/models/product.rb [hunk 13]
```diff
@@ -1,17 +1,32 @@
 class Product < ApplicationRecord
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+
+  validates_with PriceValidator
+  validates :url, presence: true, url: true
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: let's have all presence true validations together
let's also use with_options wherever possible

File: app/models/concerns/price_validator.rb [hunk 14]
```diff
@@ -0,0 +1,8 @@
+class PriceValidator < ActiveModel::Validator
```
Comment (apekshit-tewathia) [inline app/models/concerns/price_validator.rb]: why in concern folder? where it should be

File: app/models/concerns/url_validator.rb [hunk 15]
```diff
@@ -0,0 +1,10 @@
+class UrlValidator < ActiveModel::Validator
+  def validate_each(record, attribute, value)
+    unless value =~ %r{\.(gif|jpg|png)\z}i
```
Comment (apekshit-tewathia) [inline app/models/concerns/url_validator.rb]: regex in a constant

File: app/models/concerns/price_validator.rb [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+  def validate(record)
+    unless record.price > record.discount_price
```
Comment (apekshit-tewathia) [inline app/models/concerns/price_validator.rb]: what would happen if record does not have a price

Comment (apekshit-tewathia) [reply app/models/concerns/price_validator.rb]: or a discount_price

File: db/migrate/20230405115518_add_line_items_count_to_cart.rb [hunk 17]
```diff
@@ -1,6 +1,5 @@
 class AddLineItemsCountToCart < ActiveRecord::Migration[7.0]
   def change
-    add_column :carts, :line_items_count, :string
-    change_column_default :carts, :line_items_count, from: nil, to: 0
+    add_column :carts, :line_items_count, :string, default: 0
```
Comment (apekshit-tewathia) [inline db/migrate/20230405115518_add_line_items_count_to_cart.rb]: why this change here

File: db/migrate/20230405140520_change_imageurl_from_products.rb [hunk 18]
```diff
@@ -1,5 +1,13 @@
 class ChangeImageurlFromProducts < ActiveRecord::Migration[7.0]
```
Comment (apekshit-tewathia) [inline db/migrate/20230405140520_change_imageurl_from_products.rb]: why is this here

**Review by apekshit-tewathia — CHANGES_REQUESTED** [review]

File: app/models/product.rb [hunk 19]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
   has_many :line_items
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: true
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :permalink, uniqueness: true, format: { with: PERMALINK_REGEX}
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: add space before `}`

File: app/models/product.rb [hunk 20]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
...
+  validates :description, format: {with: DESCRIPTION_REGEX}
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: add space around operators

File: app/models/product.rb [hunk 21]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
...
 
-  before_destroy :ensure_not_referenced_by_any_line_item
+  with_options allow_blank: true do
+    validates_with PriceValidator
+    validates :image_url, format: { with: IMAGE_URL_REGEX, message: 'must be a URL for GIF, JPG or PNG image.'}
+    validates :price, numericality: { greater_than_or_equal_to: 0.01 }
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: let's add 0.01 in  a constant

File: app/models/user.rb [hunk 22]
```diff
@@ -2,8 +2,8 @@ class User < ApplicationRecord
   has_secure_password
 
   validates :name, presence: true, uniqueness: true
-
-  after_destroy :ensure_an_admin_remains
+  validates :email, uniqueness: true, format: {with: URI::MailTo::EMAIL_REGEXP}
```
Comment (apekshit-tewathia) [inline app/models/user.rb]: why format validation twice?

**Review by apekshit-tewathia — CHANGES_REQUESTED** [review]

File: app/models/product.rb [hunk 23]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
...
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: true
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :permalink, uniqueness: true, format: { with: PERMALINK_REGEX }
+  validates :description, format: { with: DESCRIPTION_REGEX }
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: why isnt this in allow_blank block?

File: app/models/product.rb [hunk 24]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
   has_many :line_items
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+  validates :url, presence: true, url: true
+  validates :title, :description, presence: true
   validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  validates :permalink, uniqueness: true, format: { with: PERMALINK_REGEX }
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: why isnt there a presence validation for permalink?

**Review by apekshit-tewathia — CHANGES_REQUESTED** [review]

File: app/models/product.rb [hunk 25]
```diff
@@ -2,20 +2,26 @@ class Product < ApplicationRecord
   has_many :line_items
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
+  validates :url, presence: true, url: true
+  validates :title, :description, :permalink, presence: true
   validates :title, uniqueness: true
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: this should also be in with_options block

File: app/models/line_item.rb [hunk 26]
```diff
@@ -3,6 +3,8 @@ class LineItem < ApplicationRecord
   belongs_to :product
   belongs_to :cart, optional: true
   
+  validates :product_id, uniqueness: {scope: :cart_id}
```
Comment (apekshit-tewathia) [inline app/models/line_item.rb]: add spacing around operators

File: app/models/product.rb (hunk 13 above)
Comment (apekshit-tewathia) [reply app/models/product.rb]: why is this this comment not done?

File: app/models/line_item.rb (hunk 26 above)
Comment (rachitchaurasia99) [reply app/models/line_item.rb]: done.

File: app/models/product.rb (hunk 25 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done.

File: app/models/product.rb (hunk 24 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 10 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 11 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done.

File: app/models/user.rb (hunk 22 above)
Comment (rachitchaurasia99) [reply app/models/user.rb]: fixed

File: app/controllers/concerns/url_validator.rb (hunk 8 above)
Comment (rachitchaurasia99) [reply app/controllers/concerns/url_validator.rb]: ok sir

File: app/models/product.rb (hunk 12 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: corrected

File: app/models/concerns/price_validator.rb (hunk 14 above)
Comment (rachitchaurasia99) [reply app/models/concerns/price_validator.rb]: it should be in validators folder as it makes more sense.

File: app/models/concerns/url_validator.rb (hunk 15 above)
Comment (rachitchaurasia99) [reply app/models/concerns/url_validator.rb]: done

File: app/models/product.rb (hunk 13 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: Sorry sir, will make a habit to reply on every comment.

File: app/models/concerns/price_validator.rb (hunk 16 above)
Comment (rachitchaurasia99) [reply app/models/concerns/price_validator.rb]: done

File: app/models/product.rb (hunk 23 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: corrected

File: app/models/product.rb (hunk 21 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 20 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb (hunk 19 above)
Comment (rachitchaurasia99) [reply app/models/product.rb]: done

File: app/models/product.rb [hunk 27]
```diff
@@ -2,20 +2,28 @@ class Product < ApplicationRecord
...
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
-  validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  with_options presence: true do
+    validates :url, url: true
+    validates :title, :description, :permalink
+    validates :title, uniqueness: true
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: does this not add presence validation twice on title?
Please read about what with_options does

File: rails-depot
Comment (apekshit-tewathia) [inline rails-depot:1]: what is this?

File: app/models/product.rb [hunk 28]
```diff
@@ -2,20 +2,28 @@ class Product < ApplicationRecord
...
   has_many :orders, through: :line_items
 
-  validates :title, :description, :image_url, presence: { message: "can't be blank"}
-  validates :title, uniqueness: true
-  validates :image_url, allow_blank: true, format: {with: %r{\.(gif|jpg|png)\z}i, message: 'must be a URL for GIF, JPG or PNG image.'}
-  validates :price, numericality: { greater_than_or_equal_to: 0.01 }
+  with_options presence: true do
+    validates :url, url: true
+    validates :description, :permalink
+    validates :title, uniqueness: true
```
Comment (apekshit-tewathia) [inline app/models/product.rb]: have all presence true at the top, and then  move other two(L6 and L8 in the allow_blank with_options block)

File: app/models/user.rb [hunk 29]
```diff
@@ -1,10 +1,12 @@
 class User < ApplicationRecord
   has_secure_password
 
-  validates :name, presence: true, uniqueness: true
+  with_options presence: true do
+    validates :name
+    validates :email, allow_blank: true, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (apekshit-tewathia) [inline app/models/user.rb]: does it make sense to have both presence true and allow_blank true in the same validation?
Please follow what you did in product.rb everywhere
